### ISSUE TITLE
Few initial requests failing with BufferOverflowException

### ISSUE THREAD

**Issue by kstokoz** [akka/akka-http#2727]
Hi Akka Http Team,

Using legacy pool and latest akka-http `10.1.10` few initial requests failing with `BufferOverflowException`

```
import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.model._
import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.server.Route
import akka.stream.{ActorMaterializer, ActorMaterializerSettings}
import akka.util.ByteString
import com.typesafe.config.ConfigFactory

import scala.concurrent.Await
import scala.concurrent.duration.Duration

object AppConfig {
  val actorSystemConfig = ConfigFactory.parseString(
    """
      akka {
        loglevel = DEBUG
        loggers = ["akka.event.slf4j.Slf4jLogger"]
        http {
          host-connection-pool {
           pool-implementation = legacy
           max-open-requests = 4096
           max-connections = 2048
         }
       }
     }
    """
  )
}

object ReproducerApp extends App {

  implicit val system = ActorSystem("actor-system", AppConfig.actorSystemConfig)
  implicit val materializer = ActorMaterializer(ActorMaterializerSettings.create(system).withDebugLogging(true))(system)

  val route: Route =
    path("test") {
      complete(StatusCodes.OK)
    }

  val bindingFuture = Http().bindAndHandle(route, "127.0.0.1", 8080)
  System.in.read()
}

object ReproducerClientApp extends App {
  implicit val system = ActorSystem("actor-system", AppConfig.actorSystemConfig)
  implicit val materializer = ActorMaterializer(ActorMaterializerSettings.create(system).withDebugLogging(true))(system)
  val request =
    HttpRequest(
      method = HttpMethods.PUT,
      uri    = "http://127.0.0.1:8080/test",
      entity = HttpEntity(ContentTypes.`application/json`, ByteString("test"))
    )

  (1 to 5).foreach { iindex =>
    Await.ready(Http()(system).singleRequest(request), Duration.Inf)
      .andThen {
        case result => println(s"completed $iindex with $result")
      }(system.dispatcher)
  }
}
```

Output:
```
completed 1 with Failure(akka.stream.BufferOverflowException: Exceeded configured max-open-requests value of [4096]. This means that the request queue of this pool (HostConnectionPoolSetup(127.0.0.1,8080,ConnectionPoolSetup(ConnectionPoolSettings(2048,0,5,4096,1,Duration.Inf,100 milliseconds,2 minutes,30 seconds,ClientConnectionSettings(Some(User-Agent: akka-http/10.1.10),10 seconds,1 minute,512,None,WebSocketSettings(<function0>,ping,Duration.Inf,akka.http.impl.settings.WebSocketSettingsImpl$$$Lambda$334/0x0000000800409840@2a2aa189),List(),ParserSettings(2048,16,64,64,8192,64,8388608,8388608,256,1048576,Strict,RFC6265,true,Set(),Full,Error,HashMap(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, If-None-Match -> 0, User-Agent -> 32, Content-MD5 -> 0, Date -> 0, If-Match -> 0),false,true,akka.util.ConstantFun$$$Lambda$324/0x0000000800405040@32338a0f,akka.util.ConstantFun$$$Lambda$324/0x0000000800405040@32338a0f,akka.util.ConstantFun$$$Lambda$325/0x0000000800405840@14a5c853),None,TCPTransport),Legacy,1 second),akka.http.scaladsl.HttpConnectionContext$@6bfc05a7,akka.event.MarkerLoggingAdapter@5a41a491))) has completely filled up because the pool currently does not process requests fast enough to handle the incoming request load. Please retry the request later. See http://doc.akka.io/docs/akka-http/current/scala/http/client-side/pool-overflow.html for more information.)
completed 2 with Failure(akka.stream.BufferOverflowException: Exceeded configured max-open-requests value of [4096]. This means that the request queue of this pool (HostConnectionPoolSetup(127.0.0.1,8080,ConnectionPoolSetup(ConnectionPoolSettings(2048,0,5,4096,1,Duration.Inf,100 milliseconds,2 minutes,30 seconds,ClientConnectionSettings(Some(User-Agent: akka-http/10.1.10),10 seconds,1 minute,512,None,WebSocketSettings(<function0>,ping,Duration.Inf,akka.http.impl.settings.WebSocketSettingsImpl$$$Lambda$334/0x0000000800409840@2a2aa189),List(),ParserSettings(2048,16,64,64,8192,64,8388608,8388608,256,1048576,Strict,RFC6265,true,Set(),Full,Error,HashMap(If-Range -> 0, If-Modified-Since -> 0, If-Unmodified-Since -> 0, default -> 12, If-None-Match -> 0, User-Agent -> 32, Content-MD5 -> 0, Date -> 0, If-Match -> 0),false,true,akka.util.ConstantFun$$$Lambda$324/0x0000000800405040@32338a0f,akka.util.ConstantFun$$$Lambda$324/0x0000000800405040@32338a0f,akka.util.ConstantFun$$$Lambda$325/0x0000000800405840@14a5c853),None,TCPTransport),Legacy,1 second),akka.http.scaladsl.HttpConnectionContext$@6bfc05a7,akka.event.MarkerLoggingAdapter@5a41a491))) has completely filled up because the pool currently does not process requests fast enough to handle the incoming request load. Please retry the request later. See http://doc.akka.io/docs/akka-http/current/scala/http/client-side/pool-overflow.html for more information.)
completed 3 with Success(HttpResponse(200 OK,List(Server: akka-http/10.1.10, Date: Mon, 23 Sep 2019 09:14:42 GMT),HttpEntity.Strict(text/plain; charset=UTF-8,OK),HttpProtocol(HTTP/1.1)))
completed 4 with Success(HttpResponse(200 OK,List(Server: akka-http/10.1.10, Date: Mon, 23 Sep 2019 09:14:42 GMT),HttpEntity.Strict(text/plain; charset=UTF-8,OK),HttpProtocol(HTTP/1.1)))
completed 5 with Success(HttpResponse(200 OK,List(Server: akka-http/10.1.10, Date: Mon, 23 Sep 2019 09:14:42 GMT),HttpEntity.Strict(text/plain; charset=UTF-8,OK),HttpProtocol(HTTP/1.1)))
```

**Comment by jrudolph** [comment]
I can see how that might happen in 10.0.10. If stream setup is slow,
especially after just having started up the ActorSystem, you might be able
to queue some requests before the buffer stage is ready to swallow them.

Some implementation details: There's some inherent race condition that the
buffer stage in the new `PoolInterface` is not requesting fast enough to
avoid some legitimate requests running into the limit. The reasoning was
that it will not happen because requests are scheduled using async
callbacks and the connection to the buffer is not, so the buffer requesting
from the PoolInterface would usually have precedence over processing more
incoming requests. Especially at startup that assumption doesn't seem to be
true because requests will already be queued before the PoolInterface is
running and then requests are pumped into it faster than the buffer can
react.

**Comment by kstokoz** [comment]
Hi @jrudolph 

We were using legacy pool up to v 10.1.8 and never saw issues like this, only after updating to 10.1.10 our tests started failing (we skipped 10.1.9). You are right it happening usually after start of actor system.

At the moment we moved to new pool impl and we do not observe any failures.

**Comment by jrudolph** [comment]
Moving to 10.2.0. It doesn't seem urgent to fix this since it only happens with the legacy pool. For 10.2.0, we are also planning #2120 which might solve this issue automatically.

**Comment by jrudolph** [comment]
Since that only happened with the legacy pool, which has been removed in 10.2, I'm closing this as won't fix for now.
